### DISCUSSION TITLE
ImageSharp Resize is 2x faster on Windows but 2x slower on Sandy Bridge + Linux compared to System.Drawing library

### DISCUSSION THREAD

**Discussion by mirshokraie** [SixLabors/ImageSharp#2003] [Ideas] [answered]
I'm running resize method on an image with the following code using ImageSharp:
`srcImage.Mutate(ctx => ctx.Resize(newSize.Width, newSize.Height, KnownResamplers.Bicubic, false));`
The above resize runs 2x faster on Windows compared to System.Drawing library. 
I use the following code in System.Drawing:
```
var newImage = new Bitmap(newSize.Width, newSize.Height, pixelFormat);
using (Graphics gr = Graphics.FromImage(newImage)) {
  gr.InterpolationMode = InterpolationMode.HighQualityBicubic;
  gr.PixelOffsetMode = PixelOffsetMode.HighQuality;
  gr.SmoothingMode = SmoothingMode.AntiAlias;
  gr.DrawImage(srcImage, new Rectangle(0, 0, newSize.Width, newSize.Height), new Rectangle(0,0,srcImage.Width,srcImage.Height), GraphicsUnit.Pixel); 
}
```

But when I run the same resize code using the same image on Linux, ImageSharp gets 2x slower compared to System.Drawing. The output quality in all cases look the same to me, and it also doesn't matter which image or what newSize I'm using. 
I also checked the followings and they all returned true when running in Linux: sse42, sse41, sse3, sse2, Hardware Accelerated.
I'm wondering what could be the difference that same code, same image can run so much faster on Windows and a lot slower on linux compared to System.Drawing.

BTW thanks for such an amazing library! Since our code is executed under linux, we are currently in the process of transitioning from System.Drawing library to ImageSharp...

**Comment by JimBobSquarePants** [comment]
You need to provide a benchmark when making these kinds of claims.

**Reply by mirshokraie** [reply]
Do you want me to provide a command line project with the above code or just attaching the image I used and a table of running times?

**Reply by antonfirsov** [reply]
It's about the benchmarking method / code. Naive methods can be misleading, eg. if you run the method only once, the majority of the time you see is going to be the JIT time. The [BenchMarkDotNet](https://github.com/dotnet/BenchmarkDotNet) library can take care of these kind of problems for you so you get representative numbers.

**Comment by antonfirsov** [comment]
Just did a run of our parallel, large image [LoadResizeSave benchmarks](https://github.com/SixLabors/ImageSharp/tree/main/tests/ImageSharp.Benchmarks/LoadResizeSave) and got these results on my Linux VM:

```
BenchmarkDotNet=v0.13.0, OS=ubuntu 20.04
Intel Core i9-10900X CPU 3.70GHz, 1 CPU, 10 logical and 5 physical cores
.NET SDK=5.0.201
  [Host]   : .NET Core 3.1.17 (CoreCLR 4.700.21.31506, CoreFX 4.700.21.31502), X64 RyuJIT
  ShortRun : .NET Core 3.1.17 (CoreCLR 4.700.21.31506, CoreFX 4.700.21.31502), X64 RyuJIT

Job=ShortRun  IterationCount=3  LaunchCount=1
WarmupCount=3

|        Method | maxDegreeOfParallelism |       Mean |      Error |   StdDev | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------- |----------------------- |-----------:|-----------:|---------:|------:|--------:|------:|------:|------:|----------:|
| SystemDrawing |                     10 |   538.0 ms |   303.1 ms | 16.61 ms |  1.49 |    0.09 |     - |     - |     - |     17 KB |
|    ImageSharp |                     10 |   362.3 ms |   263.1 ms | 14.42 ms |  1.00 |    0.00 |     - |     - |     - |    890 KB |
|        Magick |                     10 | 1,357.7 ms | 1,199.3 ms | 65.74 ms |  3.76 |    0.33 |     - |     - |     - |     54 KB |
|       NetVips |                     10 |   188.6 ms |   254.1 ms | 13.93 ms |  0.52 |    0.03 |     - |     - |     - |     53 KB |
```

**Reply by mirshokraie** [reply]
Thanks for the info. This is the result I'm getting from loading the image, resizing, and then saving:
BenchmarkDotNet=v0.13.1, OS=centos 7
Intel Xeon E312xx (Sandy Bridge), 1 CPU, 4 logical and 4 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
|              Method |    Mean |    Error |   StdDev |  Median |
|-------------------- |--------:|---------:|---------:|--------:|
|    ImagesharpResize | 3.688 s | 0.0733 s | 0.1931 s | 3.585 s |
| SystemDrawingResize | 1.232 s | 0.0232 s | 0.0217 s | 1.237 s |




and this is the result when I remove load and save from the benchmark and just do the resize:

BenchmarkDotNet=v0.13.1, OS=centos 7
Intel Xeon E312xx (Sandy Bridge), 1 CPU, 4 logical and 4 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
|              Method |     Mean |   Error |  StdDev |
|-------------------- |---------:|--------:|--------:|
|    ImagesharpResize | 110.6 ms | 1.08 ms | 0.90 ms |
| SystemDrawingResize | 355.6 ms | 5.45 ms | 5.09 ms |


The image I'm using is 11MB (how can I attach it here? it doesn't let me to attach because it's more than 10MB)


```
public class Resize
{
    public Image<Rgb24> imageSharpSrcImage;
    public Bitmap systemDrawingSrcImage;
    public int newSizeWidth = 10953;
    public int newSizeHeight = 8137;
    public IResampler reSampler = KnownResamplers.Bicubic;

    public Resize()
    {

    }

    [Benchmark]
    public void ImagesharpResize()
    {
        imageSharpSrcImage = Image.Load<Rgb24>("foto_no_exif_8.jpg");
        imageSharpSrcImage.Mutate(ctx => ctx.Resize(
            newSizeWidth,
            newSizeHeight,
            reSampler,
            false));
        MemoryStream memoryStream = new MemoryStream();
        IImageEncoder imageEncoder = new JpegEncoder() { Quality = 75 };
        imageSharpSrcImage.Save(memoryStream, imageEncoder);
        

    }

    [Benchmark]
    public void SystemDrawingResize()
    {
        systemDrawingSrcImage = new Bitmap("foto_no_exif_8.jpg");
        var newImage = new Bitmap(newSizeWidth, newSizeHeight, PixelFormat.Format24bppRgb);
        using (Graphics gr = Graphics.FromImage(newImage))
        {

            gr.InterpolationMode = InterpolationMode.HighQualityBicubic;
            gr.PixelOffsetMode = PixelOffsetMode.HighQuality;
            gr.SmoothingMode = SmoothingMode.AntiAlias;

            gr.DrawImage(systemDrawingSrcImage, new System.Drawing.Rectangle(0, 0, newSizeWidth, newSizeHeight), 
                new System.Drawing.Rectangle(0,0,systemDrawingSrcImage.Width,systemDrawingSrcImage.Height), GraphicsUnit.Pixel);
            MemoryStream ms = new MemoryStream();
            newImage.Save(ms, ImageFormat.Jpeg);
        }

    }
}
```

**Reply by antonfirsov** [reply]
Yes, the image is important. Maybe you can upload it to some temporary file share service (google found this: https://sharefast.me/), then I will grab it before it disappears.

Not sure how much would it affect the results, but you shoul dispose `imageSharpSrcImage` and `systemDrawingSrcImage` at the end of benchmarks, and try to reuse `MemoryStream` between the runs.

**Reply by mirshokraie** [reply]
The image should be here: https://sharefast.me/AL

**Reply by mirshokraie** [reply]
BTW I'm amazed by the speed of response from ImageSharp developers! really appreciate it!

**Reply by antonfirsov** [reply]
Thanks! Don't take it as granted though, I'm just having an insomniac night :)

@JimBobSquarePants prefers to keep away questions from GH discussions so we can direct users towards licensed support instead. I see in your bio that you come from academia, so I assume you need this for an academic project, which for me is a good enough cause to make exceptions. If I'm mistaken, persuade your employer to get a support license :)

**Reply by antonfirsov** [reply]
Just had a look at the image. Looks like you are upscaling a large image into a huge one. That's a different use case than what we are looking at in our benchmarks, so there is a chance we have some hidden issue.

I can take a look tomorrow, pretty busy today. If you want to help getting to a conclusion faster: can you benchmark (1) Decoding only, Windows VS Linux (2)  encoding of the 10953 x 8137 image only, Windows VS Linux?

**Reply by mirshokraie** [reply]
Sure I'll do that and post the results here.

**Reply by mirshokraie** [reply]
Here is the result:

Loading image on Linux:
BenchmarkDotNet=v0.13.1, OS=centos 7
Intel Xeon E312xx (Sandy Bridge), 1 CPU, 4 logical and 4 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT


|            Method |     Mean |    Error |   StdDev |   Median |
|------------------ |---------:|---------:|---------:|---------:|
|    ImagesharpLoad | 589.8 ms | 41.25 ms | 114.3 ms | 539.7 ms |
| SystemDrawingLoad | 326.9 ms | 42.68 ms | 121.1 ms | 263.5 ms |

Loading on Windows:
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1466 (21H2)
Intel Core i7-3770 CPU 3.40GHz (Ivy Bridge), 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT


|            Method |     Mean |   Error |  StdDev |
|------------------ |---------:|--------:|--------:|
|    ImagesharpLoad | 354.6 ms | 6.14 ms | 5.74 ms |
| SystemDrawingLoad | 153.8 ms | 3.02 ms | 3.59 ms |

Saving the final image as Jpeg on Linux:
BenchmarkDotNet=v0.13.1, OS=centos 7
Intel Xeon E312xx (Sandy Bridge), 1 CPU, 4 logical and 4 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT


|                  Method |       Mean |     Error |    StdDev |     Median |
|------------------------ |-----------:|----------:|----------:|-----------:|
|    ImagesharpSaveAsJpeg | 1,997.7 ms | 130.33 ms | 363.32 ms | 1,864.7 ms |
| SystemDrawingSaveAsJpeg |   684.8 ms |  29.01 ms |  79.91 ms |   656.6 ms |

Saving as Jpeg on Windows:
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1466 (21H2)
Intel Core i7-3770 CPU 3.40GHz (Ivy Bridge), 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT


|                  Method |       Mean |   Error |  StdDev |
|------------------------ |-----------:|--------:|--------:|
|    ImagesharpSaveAsJpeg | 1,079.5 ms | 6.01 ms | 5.33 ms |
| SystemDrawingSaveAsJpeg |   447.4 ms | 1.77 ms | 1.66 ms |

**Reply by JimBobSquarePants** [reply]
Thanks for the numbers. This makes it clear to me what is happening.

You are testing with processors that possess different feature sets.

I'm not seeing Avx2, Avx, or Fma in your description of the Linux environment  processor features, only  Sse42, Sse41, Sse3, Sse2

To my knowledge the processor for your Windows environment supports at least Avx.

_Note. I tried to look up both processors on the Intel processor specification docs but the search functionality isn't great._

We rely heavily on those features to get the most out of the convolution process. The time we are losing during decode/encode
(_System.Drawing is incredibly fast for both and I'd love to chat to the engineers responsible to see what the secret sauce is_) we are making up for on some systems during the resize process itself (_System.Drawing is quite slow there_).

As you can see here we optimize as much as we can based upon processor features. Our fallback method will also be optimized based upon features in the underlying runtime. For an operation like convolution which works per pixel this can make a huge difference in timings. 

https://github.com/SixLabors/ImageSharp/blob/f24b02f31c2147a76a8d9894674d889ab194dbbf/src/ImageSharp/Processing/Processors/Transforms/Resize/ResizeKernel.cs#L72-L162

BTW. I'm happy to support academia 😄

**Reply by mirshokraie** [reply]
Just tested Avx, Avx2, and Fma under my Linux and Windows environment and they both support only AVX (no support for Avx2 and Fma). So this can partly explain the slowness... Thanks for the quick responses and also for letting me know about BenchMarkDotNet. I was using stopwatch class to get approximate run time numbers and realized that it's not reliable at all :D

**Reply by JimBobSquarePants** [reply]
I'm thoroughly confused now. 

- Both environments support the same SIMD features. 
- Both environments are slower than System.Drawing for Load/Save with large images.
- Resize is 2x faster on Windows but 2X slower on Linux?

This does not make sense nor seem possible. There's different throughput for SIMD operations with different processors but this seems far to extreme a difference. 

Can you please show me a benchmark for resize only for both platforms?

**Reply by antonfirsov** [reply]
> Can you please show me a benchmark for resize only for both platforms?

https://github.com/SixLabors/ImageSharp/discussions/2003#discussioncomment-2185217 <- I think this answers the question more or less. It also it confirms what I remember from my benchmarking: resize is *much* less significant for the total LRS operation than decoder (and encoder in case of upscaling).

I think the *relative* difference adds up from several factors:
- Decoder: We loose AVX2 color conversion and [pixel packing falls back to scalar path](https://github.com/SixLabors/ImageSharp/blob/e54219acec10d2d2dc247ed5fd4230204a89c07b/src/ImageSharp/Common/Helpers/SimdUtils.Pack.cs#L28-L37)
- Encoder: [color conversion falls back to scalar path](https://github.com/SixLabors/ImageSharp/blob/e54219acec10d2d2dc247ed5fd4230204a89c07b/src/ImageSharp/Formats/Jpeg/Components/Encoder/RgbToYCbCrConverterVectorized.cs#L18-L28).
- System.Drawing could be faster on Linux than on Windows because libdiplus is backed by libjpeg-turbo. 

@mirshokraie unfortunately, I don't think we are going to invest much into creating missing SSE paths. Sandy Bridge CPU-s are more than a decade old, most users do have AVX2 today. (But community contribution is always welcome!)

**Reply by br3aker** [reply]
Can you please re-upload the image, link is dead atm. I want to check speeds on my machine, this speed difference is strange.

In benchmark you are re-saving jpeg with quality of 75 but there's no way to set quality of the `System.Drawing` according to the docs. Need to check what's the 'default' quality of `System.Drawing`.

**Reply by br3aker** [reply]
> unfortunately, I don't think we are going to invest much into creating missing SSE paths

This is partially fixable without sse code - `YCbCr` pixel type. It's binary 'equal' to the existing `Rgb24` so implementation should not be that complicated and it can eliminate a lot of overhead during load-resize-encode scenarions when initial and resulting colorspaces are the same.

**Reply by JimBobSquarePants** [reply]
@br3aker quality can be set with System.Drawing using encoder parameters.

https://github.com/bleroy/core-imaging-playground/blob/3da4e08af800f0b9999b48057e51072bdda58991/NetCore/LoadResizeSave.cs#L136-L142

**Reply by br3aker** [reply]
@JimBobSquarePants thanks! Yeah, that's what we are doing in benchmarks, thought the API used in benchmarks had some sort of a shortcut without 'parameter's :)

**Reply by mirshokraie** [reply]
@br3aker Here is the link to image: https://sharefast.me/KW

**Reply by mirshokraie** [reply]
@JimBobSquarePants 
Resize only on Windows:
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1466 (21H2)
Intel Core i7-3770 CPU 3.40GHz (Ivy Bridge), 1 CPU, 8 logical and 4 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT

|              Method |        Mean |     Error |    StdDev |
|-------------------- |------------:|----------:|----------:|
|    ImagesharpResize |    57.67 ms |  1.138 ms |  2.615 ms |
| SystemDrawingResize | 2,013.19 ms | 12.862 ms | 12.031 ms |

Resize only on Linux:
BenchmarkDotNet=v0.13.1, OS=centos 7
Intel Xeon E312xx (Sandy Bridge), 1 CPU, 4 logical and 4 physical cores
.NET SDK=6.0.101
  [Host]     : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT
  DefaultJob : .NET 6.0.1 (6.0.121.56705), X64 RyuJIT

|              Method |     Mean |   Error |   StdDev |
|-------------------- |---------:|--------:|---------:|
|    ImagesharpResize | 110.6 ms | 1.08 ms |  0.90 ms |
| SystemDrawingResize | 355.6 ms | 5.45 ms | 5.09 ms |

**Reply by JimBobSquarePants** [reply]
None of these benchmarks are showing the behaviour quoted at the beginning of this discussion?

**Reply by mirshokraie** [reply]
@JimBobSquarePants The title of this topic comes from the fact that I was using stopwatch to calculate the time, and as I mentioned I realized it is not accurate after using BenchMarkDotNet, but as you can see (in the benchmark results I posted) the total time (load, save, and resize) for ImageSharp is almost twice more than System.Drawing on linux. It's probably because System.Drawing becomes fast in resize operation on linux while it's too slow on Windows  (2 seconds on Windows vs 355 msec on Linux), and ImageSharp loses speed on loading and saving...

**Reply by br3aker** [reply, marked as answer]
We've encountered same difference in performance in 2021. `System.Drawing` is clearly cheating when re-encoding image without any modifications within the same format. Your example image before and after re-encoding with `systemDrawingSrcImage.Save(saveStream, ImageFormat.Jpeg);` yields **binary equal** file which is highly unlikely for a lossy format. I don't know if GDI is open source so I can't pinpoint the exact place where this happens.

To prove that, let's provide actual quality parameter to the encoder: 
```cs
using var @params = new EncoderParameters(1);
using var qualityParam = new EncoderParameter(Encoder.Quality, quality); // MUST be of type long
@params.Param[0] = qualityParam;
```

Your original image uses all-ones matrix meaning it has a quality of 100, which is actually a very bad value for the jpeg images but it's still 'correct':
![image](https://user-images.githubusercontent.com/20967409/154618035-e516abc0-0217-4431-b1c4-a328d91d6ff9.png)

**File size after re-encoding**
System.Drawing:  11057 kb
ImageSharp: 11547 kb
_original: 10992 kb_

As you can see, it's no longer 'equal' in binary. And even while ImageSharp writes approx. 4.4% more bytes - it's still faster than System.Drawing when Avx2 is present and not **that** slower if there's only avx available:

```
ImageSharp - has Avx2 and Fma
// Encoding q=100 | color=YCbCrRatio420
// Elapsed: 5743ms across 20 iterations
// Average: 287,15ms

ImageSharp - only Avx which is not used so basically a Sse pipeline
// Encoding q=100 | color=YCbCrRatio420
// Elapsed: 6948ms across 20 iterations
// Average: 347,4ms

System.Drawing - explicit quality parameter
// Encoding q=100
// Elapsed: 6285ms across 20 iterations
// Average: 314,25ms

System.Drawing - no parameters at all, looks fishy 
// Decoding
// Elapsed: 2460ms across 20 iterations
// Average: 123ms
```

Can't say the same for the decoder - needs further research. Maybe it's an another speed hack, maybe it's simply faster - who knows. That's why full decode-resize-encode benchmarks matter - that way we're sure `System.Drawing` doesn't do some magic with caching or whatever :)

**Reply by mirshokraie** [reply]
I think the speed mystery is finally solved! Thank you all! :)

**Comment by tocsoft** [comment]
My theory here is that System.Drawing on linux isn't actually doing the same operation as we are or windows does.
That issue implies that `InterpolationMode` is being ignored and thus its likely not doing as much work.

This issue in libgdiplus looks like it could be related https://github.com/mono/libgdiplus/issues/730

**Reply by mirshokraie** [reply]
@tocsoft That makes sense. I think this can explain why system.drawing suddenly becomes fast on linux... then I guess this kind of comparison between windows and linux may not be valid. I'll try to deploy the application on a newer machine with support for avx2 and fma and hopefully I can get a much better results on linux :)
